galley/pkg/config/analysis/analyzers/auth/mtls/policy_checker.go

```diff
@@ -0,0 +1,197 @@
+	meshHasStrictMTLSPolicy bool
+
+	namespaceToMTLSMode map[string]strictMTLSMode
+	workloadToMTLSMode  map[Workload]strictMTLSMode
+}
+
+// strictMTLSMode is a helper type used to represent whether or not MTLS was
+// explicitly enabled, explicitly disabled, or just not present. Useful for the
+// IsServiceMTLSEnforced check to determine which policy applies.
+type strictMTLSMode int
+
+const (
+	_ strictMTLSMode = iota 
+	strictMTLSExplicitlyEnabled
+	strictMTLSExplicitlyDisabled
+)
+
+// NewPolicyChecker creates a new PolicyChecker instance.
+func NewPolicyChecker() *PolicyChecker {
+	return &PolicyChecker{
+		namespaceToMTLSMode: make(map[string]strictMTLSMode),
+		workloadToMTLSMode:  make(map[Workload]strictMTLSMode),
+	}
+}
+
+// AddMeshPolicy adds a mesh-level policy to the checker. Note that there can
+// only be at most one mesh level policy in effect.
+func (pc *PolicyChecker) AddMeshPolicy(p *v1alpha1.Policy) {
+	pc.meshHasStrictMTLSPolicy = doesPolicyEnforceMTLS(p)
+}
+
+// AddPolicy adds a new policy object to the PolicyChecker to use when later
+// determining if a service is MTLS-enforced. The namespace of the policy is
+// also provided as some policies can target the local namespace.
+func (pc *PolicyChecker) AddPolicy(namespace string, p *v1alpha1.Policy) error {
+	var policyMode strictMTLSMode
+	if doesPolicyEnforceMTLS(p) {
+		policyMode = strictMTLSExplicitlyEnabled
+	} else {
+		policyMode = strictMTLSExplicitlyDisabled
+	}
+
+	if len(p.Targets) == 0 {
+		// Rule targets the namespace.
+		pc.namespaceToMTLSMode[namespace] = policyMode
+		return nil
+	}
+	// Discover the targeted workload and take note. Should normalize.
+	for _, target := range p.Targets {
+		fqdn := util.ConvertHostToFQDN(namespace, target.Name)
+
+		if len(target.Ports) == 0 {
+			// Policy targets all ports on workload
+			pc.workloadToMTLSMode[NewWorkload(fqdn)] = policyMode
+		}
+
+		for _, port := range target.Ports {
+			if port.GetName() != "" {
+				pc.workloadToMTLSMode[NewWorkloadWithPortName(fqdn, port.GetName())] = policyMode
+			} else if port.GetNumber() != 0 {
+				pc.workloadToMTLSMode[NewWorkloadWithPortNumber(fqdn, port.GetNumber())] = policyMode
+			} else {
+				// Unhandled case!
+				return fmt.Errorf("policy has a port with no name/number for target %s", target.Name)
+			}
+		}
+	}
+
+	return nil
+}
+
+// IsServiceMTLSEnforced returns true if a workload requires incoming
+// connections to use MTLS, or false if MTLS is not a hard-requirement (e.g.
+// mode is permissive, peerIsOptional is true, etc). Only call this after adding
+// all policy resources in effect via AddPolicy or AddMeshPolicy.
+func (pc *PolicyChecker) IsServiceMTLSEnforced(w Workload) (bool, error) {
+	// TODO support understanding port name -> port number mappings
+	switch pc.workloadToMTLSMode[w] {
+	case strictMTLSExplicitlyEnabled:
+		return true, nil
+	case strictMTLSExplicitlyDisabled:
+		return false, nil
+	}
```

Comment:
Do you also need to handle the "unset" case for this and the following switches?

Reply:
No, because we want the unset case to fall through. This does look confusing though, so I added an explicit case + comment to call that out (as well as a default case that errors if an unknown mtls mode is specified).